### DISCUSSION TITLE
安卓加载1280*720pag动效文件时，native增加55M左右，期望得到优化建议

### DISCUSSION THREAD

**Discussion by demo-wang-vip** [Tencent/libpag#2044] [Q&A] [answered]
【版本信息】
PAG SDK 版本：4.2.61
操作系统版本：安卓4.4，1G内存（机顶盒）

【平台信息】
Android 原生 

【问题&疑问】
我们使用pagimageview(未开启硬件加速)加载一个1280x720的动效文件，2秒左右，发现native内存会增加55M左右（测试时关闭了磁盘缓存，如果不关闭，第一次占用55M左右，后续使用缓存，内存是正常的）
当前测试的屏幕分辨率为1920x1080，按照单帧图片渲染的内存计算，1920x1080x4[byte]，约为8M，再加上其他内存占用，预期应该不超过20M，与55M的实际情况相差比较大。
请问有什么方式能够优化，从而降低渲染的内存吗？

**Comment by kevingpqi123** [comment, marked as answer]
PAGImageView 首次渲染的时候，需要渲染出结果构建出磁盘缓存，渲染的时候 OpenGL 渲染环境、pag 内部渲染缓存都需要占用内存，如果素材文件中含有 BMP 预合成（视频序列帧），创建硬件解码器也需要占用内存，解码过程中也要缓存几帧视频数据同样需要占用内存，这些内存占用是不可或缺的，当渲染过一次后，全部渲染结果会缓存到磁盘，渲染环境相关的内存都会释放掉，后续仅仅是读取渲染磁盘文件的开销，内存占用比较小
如果想降低首次内存：
1、首先是优化素材，尽可能的采用纯矢量的素材；
2、是否可以减小绘制面积
主要是这两个方面的优化
